### PR TITLE
Observers refactor

### PR DESCRIPTION
**Pull request by andreafioraldi**
This PR introduces the HasExecHooks trait for pre_exec and post_exec hooks.
These hooks take the state, the manager and the input as parameters.
Now the hooks are separated from the definition of the Observer trait, otherwise we need to introduce EM, S and I as generics to observers and then feedbacks.
Executors should implement HasExecHooks too, that is now a common interface.

### CODE REVIEW

File: libafl/src/lib.rs [hunk 1]
```diff
@@ -194,7 +194,6 @@ mod tests {
 
         let mut harness = |_buf: &[u8]| ExitKind::Ok;
         let mut executor = InProcessExecutor::new(
-            "main",
```
Comment (domenukk) [inline libafl/src/lib.rs:197]: Names for executors are no longer en-vogue?

**Comment by domenukk** [comment]
Looks good to me, @s1341 maybe you want to look over the frida changes ;)

File: libafl/src/lib.rs (hunk 1 above)
Comment (andreafioraldi) [reply libafl/src/lib.rs:197]: not needed, we can support multiple executors in another way

**Comment by andreafioraldi** [comment]
> Looks good to me, @s1341 maybe you want to look over the frida changes ;)

He can look directly in merge, I need to swuash to start working on #24

**Comment by s1341** [comment]
I'll review later when I merge dev back into launcher...

**Pull request merged**